FEATURE-User-Functionality

squarecylinder opened the pull request:
Need to have users
Will call this PR done 
- [ ] When a user can sign up(Maybe do some email auth?)
- [ ] When a user can login
- [ ] When users can bookmark stories, meaning they it gets saved to the database as a story the user follows and shows up on their account page. I would like it if we can show a badge or something if stories that are bookmarked get updated.
- [ ] When users can upload profile pictures
- [ ] When users can give a rating to a story. It doesn't have to be one time but it should be persistent, so if a user gave a story a 5 but a few arcs later gives it a 3 the story should have only the latest rating from the user.
- [ ] When a user can comment on a story or chapter
- [ ] When users can log out
- [ ] When we can display in someway that a user has already read a chapter, ie graying out the title or something like that on the story details page

Commit:
fails

Commit:
Users can now be saved to database!

Commit:
User can now login

Commit:
User signup/login/jwt

Commit:
File Structure Clean Up

Commit:
remove package-lock

Glen-Cocoa commented on StoryTowerWeb/package-lock.json:
```diff
@@ -1935,10 +1941,40 @@
+      "integrity": "sha512-oSXzaWypCMHkPC3NvBEaPHf0KsA5mvPrOPgQWDsbg8n7orZ290M0BmC/jgRZ4vcJ6DTAhjrsSYgdsW/F+MFOBA=="
+    },
+    "node_modules/lodash.isstring": {
+      "version": "4.0.1",
+      "resolved": "https://registry.npmjs.org/lodash.isstring/-/lodash.isstring-4.0.1.tgz",
+      "integrity": "sha512-0wJxfxH1wgO3GrbuP+dTTk7op+6L41QCXbGINEmD+ny/G/eCqGzxyCsh7159S+mgDDcoarnBw6PC1PS5+wUGgw=="
+    },
+    "node_modules/lodash.once": {
+      "version": "4.1.1",
+      "resolved": "https://registry.npmjs.org/lodash.once/-/lodash.once-4.1.1.tgz",
+      "integrity": "sha512-Sb487aTOCr9drQVL8pIxOzVhafOjZN9UU54hiN8PU3uAiSV7lx1yYNpbNmex2PK6dSJoNTSJUUswT651yww3Mg=="
```
DELETE THUIS

Commit:
need default bookmark values, help!

Glen-Cocoa commented on StoryTowerWeb/schemas/mutations.js:
```diff
@@ -1,4 +1,7 @@
+const { signToken } = require('../utils/auth.js');
 
 const addScrapedDataToCatalog = async ({ scrapedData }) => {
   console.log('GraphQL mutation for adding scraped data to catalog is executed!');
```
should this be a `console.debug`?

https://developer.mozilla.org/en-US/docs/Web/API/console/debug

squarecylinder replied:
We can try that!

Glen-Cocoa commented on StoryTowerWeb/schemas/mutations.js:
```diff
@@ -29,6 +32,52 @@ const addScrapedDataToCatalog = async ({ scrapedData }) => {
 
+const createUser = async (_, { email, username, password }) => {
+  try {
+    const existingUser = await User.findOne({ $or: [{ email }, { username }] });
```
maybe name this something else? like `userExists`

just reads a little more naturally

squarecylinder replied:
I think since this is finding an object and not returning a bool it makes sense as is. But I have no preference on naming

Glen-Cocoa commented on StoryTowerWeb/schemas/mutations.js:
```diff
@@ -29,6 +32,52 @@ const addScrapedDataToCatalog = async ({ scrapedData }) => {
+  try {
+    const existingUser = await User.findOne({ $or: [{ email }, { username }] });
+    if (existingUser) {
+      throw new Error('User with this email or username already exists.');
```
do we want to throw here? won't that crash? we could `console.error` or just return an error message to UI instead

squarecylinder replied:
This just puts an error in the console. I didnt see any crashing.

Commit:
cant get them saving

Commit:
Can now pass user through context

Commit:
bookmarkedStories: [ new ObjectId("64c99c0ddfacc14d4fef3ea7") ],
This should be an object with title and coverart
I don't know why only the objectid is showing up

Commit:
I think we have bookmarks!

Commit:
biting the bullet, adding use context

Commit:
converting to context

Commit:
slowly migrating to usecontext

Glen-Cocoa commented on storytower/src/components/StoryDetails/StoryDetails.jsx:
```diff
@@ -43,14 +61,18 @@ const StoryDetails = () => {
-
+                    {isLoggedIn && (
+                        <button onClick={handleBookmarkClick}>
+                            {isBookmarked ? 'Unbookmark' : 'Bookmark'}
```
some way to 
a) make sure that the state displayed always matches the state accurately
b) prevent a user from firing a hundred db calls by spamming this button

squarecylinder replied:
Fixed the first one, not the second one. Since this PR is getting out of scope, I'm going to merge it and make another branch specifically for bookmark features.

Commit:
context refactor

Commit:
Swapped to use context
Bookmarking now works
